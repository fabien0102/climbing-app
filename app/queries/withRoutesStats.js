import { graphql } from "react-apollo";
import routesStatsQuery from "./routesStatsQuery.graphql";
import _ from "lodash";
import { startOfWeek, endOfWeek } from "date-fns";

export const getStats = routes => {
  return _(routes)
    .groupBy("grade")
    .map((r, grade) => ({
      grade,
      total: r.length,
      finished: r.filter(i => _.map(i.tries, "successLevel").includes(5))
        .length,
      flashed: r.filter(i => _.map(i.tries, "successLevel")[0] === 5).length,
      afterWork: r.filter(
        i =>
          _.map(i.tries, "successLevel").includes(5) &&
          _.map(i.tries, "successLevel")[0] !== 5
      ).length,
      notFinished: r.filter(
        i => !_.map(i.tries, "successLevel").includes(5) && i.tries.length > 0
      ).length
    }))
    .value();
};

export default graphql(routesStatsQuery, {
  options: ({ me: { id } }) => {
    const today = new Date();
    const startDate = startOfWeek(today).toISOString();
    const endDate = endOfWeek(today).toISOString();
    return {
      variables: { userId: id, startDate, endDate, null: null }
    };
  },
  props: ({ data: { refetch, routes, loading, error } }) => {
    const stats = getStats(routes);

    return {
      routesStats: {
        refetch,
        loading,
        error,
        perGrade: stats,
        total: _.sumBy(stats, "total"),
        flash: _.sumBy(stats, "flashed"),
        afterWork: _.sumBy(stats, "afterWork"),
        notFinished: _.sumBy(stats, "notFinished")
      }
    };
  }
});
