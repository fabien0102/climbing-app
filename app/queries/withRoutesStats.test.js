import { getStats } from "./withRoutesStats";
import _ from "lodash";

const { data: { routes } } = require("./__tests__/routesStatsData.json");

describe("withRoutesStats", () => {
  describe("getStats", () => {
    it("should return correct value", () => {
      expect(getStats(routes)).toMatchSnapshot();
    });

    it("should return the total", () => {
      const stats = getStats(routes);
      const total = _.sumBy(stats, "total");
      expect(total).toBe(259);
    });
  });
});
