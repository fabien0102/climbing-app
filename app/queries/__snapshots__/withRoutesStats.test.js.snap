// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withRoutesStats getStats should return correct value 1`] = `
Array [
  Object {
    "afterWork": 1,
    "finished": 5,
    "flashed": 4,
    "grade": "3",
    "notFinished": 0,
    "total": 35,
  },
  Object {
    "afterWork": 6,
    "finished": 13,
    "flashed": 7,
    "grade": "4",
    "notFinished": 0,
    "total": 50,
  },
  Object {
    "afterWork": 3,
    "finished": 7,
    "flashed": 4,
    "grade": "5",
    "notFinished": 1,
    "total": 75,
  },
  Object {
    "afterWork": 1,
    "finished": 5,
    "flashed": 4,
    "grade": "6",
    "notFinished": 1,
    "total": 71,
  },
  Object {
    "afterWork": 0,
    "finished": 0,
    "flashed": 0,
    "grade": "7",
    "notFinished": 0,
    "total": 26,
  },
  Object {
    "afterWork": 0,
    "finished": 0,
    "flashed": 0,
    "grade": "8",
    "notFinished": 0,
    "total": 2,
  },
]
`;
